### PR TITLE
Status: Online, Offline, Typing...

### PR DESCRIPTION
**Pull request by DonnaDona**

### CODE REVIEW

File: client/src/store/index.js [hunk 1]
```diff
@@ -24,6 +24,7 @@ export default new Vuex.Store({
       state.user = null;
     },
     setSocket(state, { socket }) {
+      console.log('Setting state: ', socket);
```
Comment (micheledallerive) [inline client/src/store/index.js:27]: Remove

File: client/src/views/AppContainer.vue [hunk 2]
```diff
@@ -28,17 +28,40 @@ export default {
...
+    this.$store.commit('setSocket', { socket });
 
     socket.on('chats:read', (chats) => {
       console.log('EVENT chats:read -', chats);
       this.chats = chats.map((chat) => new Chat(chat));
       // this.setActiveChat(this.chats[0]._id);
     });
 
+    socket.on('users:online', (userId) => {
+      console.log('EVENT users:online -', userId);
```
Comment (micheledallerive) [inline client/src/views/AppContainer.vue:43]: Remove

File: client/src/views/AppContainer.vue [hunk 3]
```diff
@@ -28,17 +28,40 @@ export default {
...
+        chat.members.forEach((member) => {
+          if (member._id === userId) {
+            member.online = true;
+          }
+        });
+      });
+    });
+
+    socket.on('users:offline', (userId) => {
+      console.log('EVENT users:offline -', userId);
```
Comment (micheledallerive) [inline client/src/views/AppContainer.vue:55]: Remove

File: routes/chat.js [hunk 4]
```diff
@@ -94,6 +94,13 @@ router.post('/chats', async function (req, res) {
     let members = await UserChat.find({ chat: chatId }).populate('user').exec();
     members = members.map((member) => member.user);
 
+    members = members.map((member) => {
+      return {
+        ...member._doc,
+        online: req.app.locals.onlineUsers.has(member._id.toString()),
+      };
+    });
+
```
Comment (micheledallerive) [inline routes/chat.js:103]:
```suggestion
    const onlineUsers = req.app.locals.onlineUsers;
    members = members.map((member) => {
      return {
        ...member._doc,
        online: onlineUsers.has(member._id.toString()),
      };
    });
```

File: serverSocket.js [hunk 5]
```diff
@@ -4,7 +4,8 @@ const { authMiddleware } = require('./middlewares/socket.middleware');
 const { ObjectId } = require('mongodb');
 
 // Initialize the socket.io server
```
Comment (micheledallerive) [inline serverSocket.js:6]: Update comment explaining why you are also passing `onlineUsers` as a parameter

File: serverSocket.js [hunk 6]
```diff
@@ -26,16 +27,34 @@ function init(server) {
...
           .exec();
         return { ...chat._doc, members: members.map((member) => member.user) };
       })
     );
+
+    /**
+     * chats: [ chat, chat, chat ]
+     * chat: {_id: '...', members: [user, user, user], messages: [message, message, message]}
+     * user: {_id: '...', name: '...', email: '...', password: '...', online: ?}
+     */
```
Comment (micheledallerive) [inline serverSocket.js:45]: Remove

File: serverSocket.js [hunk 7]
```diff
@@ -50,8 +69,26 @@ function init(server) {
...
     socket.on('disconnect', () => {
       console.log('⛔User disconnected with id ' + socket.id);
+
+      onlineUsers.delete(socket.userId);
+      io.emit('users:offline', socket.userId);
+    });
+
+    socket.on('user:typing', (data) => {
+      const { chatId } = data;
+      console.log(socket.userId + ' is typing in chat ' + chatId);
```
Comment (micheledallerive) [inline serverSocket.js:84]: Remove

File: serverSocket.js [hunk 8]
```diff
@@ -50,8 +69,26 @@ function init(server) {
...
+      io.to(chatId).emit('user:typing', { chatId, userId: socket.userId });
+    });
+
+    socket.on('user:notTyping', (data) => {
+      const { chatId } = data;
+      console.log(socket.userId + ' is not typing in chat ' + chatId);
```
Comment (micheledallerive) [inline serverSocket.js:90]: Remove

**Review by micheledallerive — CHANGES_REQUESTED** [review]

**Review by micheledallerive — DISMISSED** [review]

File: serverSocket.js [hunk 9]
```diff
@@ -26,16 +27,28 @@ function init(server) {
...
+      chat.members = chat.members.map((member) => {
+        return {
+          ...member._doc,
+          online: onlineUsers.has(member._id.toString()),
+        };
+      });
+      return chat;
+    });
+
+    return chats;
```
Comment (Hwid2121) [inline serverSocket.js:51]: Is this possible to make a chain of promises instead to assign to chats 2 times and make it const?

**Review by Hwid2121 — COMMENTED** [review]
Let me know if you can change that in a promise chain if not is ok! @DonnaDona

File: serverSocket.js [hunk 10]
```diff
@@ -26,16 +27,28 @@ function init(server) {
...
-    const chats = userChats;
+    let chats = userChats;
     // find all the users in each chat ad add it as a property 'members'
-    return await Promise.all(
+    chats = await Promise.all(
       chats.map(async (chat) => {
         const members = await UserChat.find({ chat: chat._id })
           .populate('user')
           .exec();
         return { ...chat._doc, members: members.map((member) => member.user) };
```
Comment (micheledallerive) [inline serverSocket.js:37]:
```suggestion
        return { ...chat._doc, members: members.map((member) => {
            return {...member.user._doc, online: onlineUsers.has(member.user._id.toString())};
        }) };
```

Comment (micheledallerive) [reply serverSocket.js:37]: Something like this could work.

File: serverSocket.js (hunk 9 above)
Comment (DonnaDona) [reply serverSocket.js]: Thanks nico!

**Comment by JekxDevil** [comment]
next features concerning the chat shall be put in a new branch named `feature/nameofthefeature`.
This branch already fulfilled its purpose beyond its completion, in fact user status could be done in a separate branch.
After every milestone, branches shall be closed for convenience and clarity.

**Pull request merged**